### PR TITLE
feat: support zero-copy with splice

### PR DESCRIPTION
**Pull request by ihciah**
- [x] Make copy more generic
- [x] Add tests

### CODE REVIEW

**Comment by ihciah** [comment]
TODO: make copy more generic.

File: monoio/src/net/unix/pipe.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    #[cfg(target_os = "linux")]
+    let flag = {
+        if crate::driver::op::non_blocking() {
+            libc::O_NONBLOCK
+        } else {
+            0
+        }
+    };
+    #[cfg(target_os = "linux")]
+    crate::syscall!(pipe2(pipes.as_mut_ptr() as *mut libc::c_int, flag))?;
```
Comment (johntaiko) [inline monoio/src/net/unix/pipe.rs:33]: Use `ptr` same as below?

File: monoio/src/io/as_fd.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+//! We impl AsReaderFd and AsWriterFd for some structs.
+
+use crate::driver::shared_fd::SharedFd;
+
+/// Get a readable shared fd from self.
+pub trait AsReaderFd {
```
Comment (johntaiko) [inline monoio/src/io/as_fd.rs:6]: Maybe need use the `AsReadFd` or `AsWriteFd` as the naming style.

Comment (dyxushuai) [inline monoio/src/io/as_fd.rs]: Maybe need use the AsReadFd or AsWriteFd as the naming style.

File: monoio/src/net/unix/pipe.rs (hunk 1 above)
Comment (dyxushuai) [inline monoio/src/net/unix/pipe.rs]: The way to use `ptr` need same as below?
```rust
crate::syscall!(pipe2(pipes.as_mut_ptr() as _, flag))?;
```

File: monoio/src/driver/op.rs [hunk 3]
```diff
@@ -106,12 +109,12 @@ impl<T> Drop for Op<T> {
 
 #[allow(unused)]
 #[cfg(not(target_os = "linux"))]
-fn non_blocking() -> bool {
+pub(crate) fn non_blocking() -> bool {
     true
```
Comment (dyxushuai) [inline monoio/src/driver/op.rs:113]: Mark function as const

File: monoio/src/driver/op.rs [hunk 4]
```diff
@@ -106,12 +109,12 @@ impl<T> Drop for Op<T> {
...
 }
 
 #[cfg(target_os = "linux")]
-fn non_blocking() -> bool {
+pub(crate) fn non_blocking() -> bool {
```
Comment (dyxushuai) [inline monoio/src/driver/op.rs:117]: Same with blew

File: monoio/Cargo.toml [hunk 5]
```diff
@@ -59,6 +59,8 @@ async-cancel = []
 # enanle zero copy(enable SOCK_ZEROCOPY + MSG_ZEROCOPY flag)
 # WARNING: this feature may cause performance degradation
 zero-copy = []
+# splice op(require kernel 5.7+)
```
Comment (dyxushuai) [inline monoio/Cargo.toml:62]: Why we need a feature flag for `splice`, maybe we can assert the kernel version.

File: monoio/src/driver/op.rs (hunk 3 above)
Comment (ihciah) [reply monoio/src/driver/op.rs:113]: Inconsistent with #[cfg(target_os = "linux")]?

File: monoio/src/driver/op.rs (hunk 4 above)
Comment (ihciah) [reply monoio/src/driver/op.rs:117]: Here can't be const

File: monoio/Cargo.toml (hunk 5 above)
Comment (ihciah) [reply monoio/Cargo.toml:62]: It will make us require higher kernel version(5.4->5.7).

Comment (ihciah) [reply monoio/Cargo.toml:62]: Maybe in the future the feature will be delete and make it as default.

**Review by dyxushuai — APPROVED** [review]

**Pull request merged**